sources/academy/webscraping/scraping_basics_python/09_getting_links.md

```diff
@@ -0,0 +1,392 @@
+    for row in data:
+        writer.writerow(row)
+
+def serialize(obj):
+    if isinstance(obj, Decimal):
+        return str(obj)
+    raise TypeError("Object not JSON serializable")
+
+with open("products.json", "w") as file:
+    json.dump(data, file, default=serialize)
+```
+
+Let's introduce several functions to make the whole thing easier to digest. First, we can turn the beginning of our program into this `download()` function, which takes a URL and returns a `BeautifulSoup` instance:
+
+```py
+def download(url):
+    response = httpx.get(url)
+    response.raise_for_status()
+
+    html_code = response.text
+    return BeautifulSoup(html_code, "html.parser")
+```
+
+Next, we can put parsing into a `parse_product()` function, which takes the product item element and returns the dictionary with data:
+
+```py
+def parse_product(product):
+    title = product.select_one(".product-item__title").text.strip()
+
+    price_text = (
+        product
+        .select_one(".price")
+        .contents[-1]
+        .strip()
+        .replace("$", "")
+        .replace(",", "")
+    )
+    if price_text.startswith("From "):
+        min_price = Decimal(price_text.removeprefix("From "))
+        price = None
+    else:
+        min_price = Decimal(price_text)
+        price = min_price
+
+    return {"title": title, "min_price": min_price, "price": price}
+```
+
+Now the CSV export. We'll make a small change here. Having to specify the field names is not ideal. What if we add more field names in the parsing function? We'd always have to remember to go and edit the export function as well. If we could figure out the field names in place, we'd remove this dependency. One way would be to infer the field names from the dictionary keys of the first row:
+
+```py
+def export_csv(file, data):
+    # highlight-next-line
+    fieldnames = list(data[0].keys())
+    writer = csv.DictWriter(file, fieldnames=fieldnames)
+    writer.writeheader()
+    for row in data:
+        writer.writerow(row)
+```
+
+:::note Fragile code
+
+The code above assumes the `data` variable contains at least one item, and that all the items have the same keys. This isn't robust and could break, but in our program, this isn't a problem, and omitting these corner cases allows us to keep the code examples more succinct.
+
+:::
+
+The last function we'll add will take care of the JSON export. For better readability of the JSON export, let's make a small change here too and set the indentation level to two spaces:
+
+```py
+def export_json(file, data):
+    def serialize(obj):
+        if isinstance(obj, Decimal):
+            return str(obj)
+        raise TypeError("Object not JSON serializable")
+
+    # highlight-next-line
+    json.dump(data, file, default=serialize, indent=2)
+```
+
+Now let's put it all together:
+
+```py
+import httpx
+from bs4 import BeautifulSoup
+from decimal import Decimal
+import csv
+import json
+
+def download(url):
+    ...
+
+def parse_product(product):
```

Comment:
Yeeah, I know it's better; I’m just struggling with the feeling that having a huge blob of code is intimidating for me and the reader 😅 But you’re absolutely right in your points. Now I can combat that feeling, wielding your comment like a two-handed melee weapon—this will help.

Reply:
maybe break up the code initially explaining what each part does, and then do something like 

>Let's put it together

and get the whole code sample as a copiable codeblock?

Reply:
@TC-MO That's what is happening, I think. But I just opted to have `...` instead of full functions in the code block, so it isn't copiable. Well, wasn't. Just pushed a commit which resolves this.